<a href="https://colab.research.google.com/github/emma-s137/cs344/blob/master/homework4.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

1. Speculate on whether you believe that so-called “deep” neural networks are destined to be another bust just as perceptrons and expert systems were in the past, or whether they really are a breakthrough that will be used for years into the future. Please give a two-to-three-paragraph answer, including examples to back up your argument.





I believe that the "deep" neural networks will not just be another bust as peceptrons or expert systems. One reason for this is because the focus a "deep" neural network is more data. In the digital, we are know living in these machine learning models will become easier to implement, as more and more data is collected. There are people in the the transhumanist movements that hope to completely monitor the activity of their own body (individual mentioned in Mark O'Connell's book To Be A Machine). Already apple watches and advancements towards smart homes (such as fridges and fish tanks) have given us more acess to data then ever before. A neural network that then learns from data seems to be an optimal direction for AI.

The failure with expert systems was the idea that the world behaves on the basis of rules. The idea that these rules can be hard coded, and that by logical reasoning an AI can discern the answer to a question. However, in many complex problems we see that certain rules will take precendence over others, and sometimes rules can be completely thrown out. It was also difficut when a new body of knowledge was discovered. The expert system would then need to be completely updated. However, with neural networks the answer is just to feed in more data. One specific problem, that proved very difficult for the rigidness of expert systems, but was more successful with neural nets was voice recognition.

The perceptron by itself proved to be to simple for solving more nuanced problems with many different parameters and patterns that could be picked up on. The perceptron was limited was a linear classified, but layering of these nodes alllowed for more complex problems to be solved.





2. Hand-compute a single, complete back-propagation cycle. Use the example network from class and compute the updated weight values for the first gradient descent iteration for the XOR example, i.e., [1, 1] → 0. Use the same initial weights we used in the class example but assume the identity function as the activation function (f(x) = x).


See image.



3. Build a Keras-based ConvNet for Keras’s Fashion MNIST dataset (fashion_mnist). Experiment with different network architectures, submit your most performant network, and report the results.

In [0]:
# References
# https://keras.io/layers/convolutional/
# https://www.tensorflow.org/api_docs/python/tf/keras/optimizers/
# https://victorzhou.com/blog/keras-neural-network-tutorial/

from keras.datasets import fashion_mnist
from keras.utils import to_categorical

(train_images, train_labels), (test_images, test_labels) = fashion_mnist.load_data()

train_images = train_images.reshape((60000, 28, 28, 1))
train_images = train_images.astype('float32') / 255

test_images = test_images.reshape((10000, 28, 28, 1))
test_images = test_images.astype('float32') / 255

train_labels = to_categorical(train_labels)
test_labels = to_categorical(test_labels)

# Split validation and trainning set with an 80, 20 plit
val_images = train_images[54000:]
train_images = train_images[:54000]

val_labels = train_labels[54000:]
train_labels = train_labels[:54000]

val_data = (val_images, val_labels)


In [91]:
from keras import layers
from keras import models

model = Sequential()
# Two Convolution layers that are 2D for pictures
model.add(layers.Conv2D(32, kernel_size=(3, 3),
                 activation='relu',
                 input_shape=(28,28,1)))
model.add(layers.Conv2D(64, (3, 3), activation='relu'))
model.add(layers.MaxPooling2D(pool_size=(2, 2)))

#Simplify output and put into 10 different categories
model.add(layers.Flatten())
model.add(layers.Dense(64, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))
# Take a look at the model summary
model.summary()


Model: "sequential_14"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
conv2d_18 (Conv2D)           (None, 26, 26, 32)        320       
_________________________________________________________________
conv2d_19 (Conv2D)           (None, 24, 24, 64)        18496     
_________________________________________________________________
max_pooling2d_9 (MaxPooling2 (None, 12, 12, 64)        0         
_________________________________________________________________
flatten_10 (Flatten)         (None, 9216)              0         
_________________________________________________________________
dense_23 (Dense)             (None, 64)                589888    
_________________________________________________________________
dense_24 (Dense)             (None, 10)                650       
Total params: 609,354
Trainable params: 609,354
Non-trainable params: 0
_______________________________________________

In [92]:
#categorical_crossentropy for making groups
model.compile(loss='categorical_crossentropy',
             optimizer='adagrad',
             metrics=['accuracy'])

model.fit(train_images,
         train_labels,
         batch_size=64,
         epochs=5,
         validation_data=val_data)


Train on 54000 samples, validate on 6000 samples
Epoch 1/5
54000/54000 [==============================] - 123s 2ms/step - loss: 0.4106 - accuracy: 0.8551 - val_loss: 0.3196 - val_accuracy: 0.8840
Epoch 2/5
54000/54000 [==============================] - 124s 2ms/step - loss: 0.2815 - accuracy: 0.8994 - val_loss: 0.2782 - val_accuracy: 0.9010
Epoch 3/5
54000/54000 [==============================] - 125s 2ms/step - loss: 0.2472 - accuracy: 0.9119 - val_loss: 0.2595 - val_accuracy: 0.9065
Epoch 4/5
54000/54000 [==============================] - 125s 2ms/step - loss: 0.2260 - accuracy: 0.9196 - val_loss: 0.2466 - val_accuracy: 0.9120
Epoch 5/5
54000/54000 [==============================] - 129s 2ms/step - loss: 0.2100 - accuracy: 0.9239 - val_loss: 0.2416 - val_accuracy: 0.9135


In [93]:
# Evaluate the model on sets
print(model.evaluate(train_images, train_labels))
print(model.evaluate(val_images, val_labels))
print(model.evaluate(test_images, test_labels))

54000/54000 [==============================] - 33s 616us/step
[0.19506046934812157, 0.9306666851043701]
6000/6000 [==============================] - 4s 628us/step
[0.24163730388879776, 0.9135000109672546]
10000/10000 [==============================] - 6s 614us/step
[0.257784378182888, 0.9065999984741211]


 
The trainning set has an accuracy of 93.1%,
   the validation set is 91.3%
   and the testing set is 90.6%.

This seems to be a good model without too much overfitting.